### PR TITLE
merge scripts

### PR DESCRIPTION
**Pull request by Cofflag**

### CODE REVIEW

File: asm/code_vcount.s [hunk 1]
```diff
@@ -4,51 +4,7 @@
...
-	pop {r4, r5, r6}
-	pop {r0}
-	bx r0
-	.align 2, 0
-_0800707C: .4byte gUnk_03003230
-_08007080: .4byte gUnk_03003360
-_08007084: .4byte gUnk_030031D0
-_08007088: .4byte 0x04000004
-
+/*
```
Comment (jiangzhengwenjz) [inline asm/code_vcount.s:7]: Why not just remove the functions?

File: include/vcount.h [hunk 2]
```diff
@@ -1,20 +1,17 @@
...
-    u16 *unk10;
-    u16 unk14;
-    u16 unk16;
-    u16 *unk18;
-    u16 unk1C;
-    u16 unk1E;
+};
+/* size = 0x8 */
+
+struct Unk_030031D0
```
Comment (jiangzhengwenjz) [inline include/vcount.h:12]: Then are you sure it's a struct, instead of a plain array?

File: src/codec_08004544.c [hunk 3]
```diff
@@ -53,3 +53,30 @@ u16 sub_08004614(u32 r0, u32 r1, u32 r2, u32 r3, u32 sp10, u32 sp14, u32 sp18)
     sp10 <<= 0x10;
     return  v1 | sp14 | sp18 | (r3 >> 8 | sp10 >> 14);
 }
+
+u32 sub_0800463C(u16 r0)
+{
+    register u32 r2 asm("r2") = gUnk_03002990[0] & (0x100 << r0); // TODO:fix fakematch
+    return (r2 | -r2) >> 0x1f;
```
Comment (jiangzhengwenjz) [inline src/codec_08004544.c:60]: Makes zero sense. 
(var | -var) >> 0x1f is just an optimization for (var != 0).

File: src/codec_08004544.c [hunk 4]
```diff
@@ -53,3 +53,30 @@ u16 sub_08004614(u32 r0, u32 r1, u32 r2, u32 r3, u32 sp10, u32 sp14, u32 sp18)
...
+}
+
+void sub_08004658(u16 r0, u16 *r1, u16 *r2, u16 *r3)
+{
+    u16 *ptr = gUnk_03002990;
+    u16 r4 = ptr[r0];
+    ptr++;
+    r4 = ptr[r0]; // TODO:fix fakematch
```
Comment (jiangzhengwenjz) [inline src/codec_08004544.c:68]: Maybe we can spot the issue if you provide a scratch on CE.

File: src/codec_08004544.c [hunk 5]
```diff
@@ -53,3 +53,30 @@ u16 sub_08004614(u32 r0, u32 r1, u32 r2, u32 r3, u32 sp10, u32 sp14, u32 sp18)
...
+    *r2 = 3 & r4;
+    *r3 = 0xc000 & r4;
+}
+
+void sub_08004684(u16 r0, u16 *r1, u16 *r2)
+{
+    u16 *ptr = gUnk_03002990;
+    u16 r3 = ptr[r0];
+    ptr++;
+    r3 = ptr[r0]; // TODO:fix fakematch
```
Comment (jiangzhengwenjz) [inline src/codec_08004544.c:79]: same here

File: src/script.c [hunk 6]
```diff
@@ -146,3 +153,152 @@ u32 sub_08012F60(u16 r0)
...
+        gUnk_030067A8[(r0 - 0x100) >> 3] &= ~(1 << r4);
+    }
+    else
+    {
+        gUnk_030067A8[(r0 - 0x100) >> 3] |= (1 << r4);
+    }
+}
+
+/*
+    Judging from these function's behavior, csm3 tends to store a signed value as unsigned,and get that value as signed.
```
Comment (jiangzhengwenjz) [inline src/script.c:185]:
```suggestion
    Judging from these function's behavior, csm3 tends to store a signed value as unsigned and get that value as signed.
```

File: src/script.c [hunk 7]
```diff
@@ -146,3 +153,152 @@ u32 sub_08012F60(u16 r0)
...
+
+u32 sub_080131DC(void)
+{
+    u16 r2 = *gUnk_03006574->unk4;
+    u32 r4;
+    gUnk_03006574->unk4++;
+    r4 = r2 << 0x10;
+    if (sub_08012578() == 0)
+    {
+        gUnk_03006574->unk4 = (r4 >> 0x11 << 1) + 0x10 + gUnk_03006578->unk28; // TODO: I think this is a fakematch.
```
Comment (jiangzhengwenjz) [inline src/script.c:263]: Again, make a scratch.

File: src/script.c [hunk 8]
```diff
@@ -146,3 +153,152 @@ u32 sub_08012F60(u16 r0)
...
+u32 sub_080132B0(void)
+{
+    return 0;
+}
+
+u32 sub_080132B4(void)
+{
+    u32 r0 = 0;
+    gUnk_030067B0[4] = r0;
+    sub_08001AEC(r0); // is this UB?
```
Comment (jiangzhengwenjz) [inline src/script.c:302]: Why?

File: src/script.c [hunk 9]
```diff
@@ -146,3 +153,152 @@ u32 sub_08012F60(u16 r0)
...
+        gUnk_03006574->unk4 = (r4 >> 0x11 << 1) + 0x10 + gUnk_03006578->unk28; // TODO: I think this is a fakematch.
+    }
+    gUnk_03006578->unk1 = 1;
+    return 0;
+}
+
+u32 sub_08013220(void)
+{
+    u16 r1 = *gUnk_03006574->unk4++;
+    gUnk_03006574->unk4 = (r1 >> 1 << 1) + 0x10 + gUnk_03006578->unk28;
```
Comment (jiangzhengwenjz) [inline src/script.c:272]: Does it work for the function above?

**Review by jiangzhengwenjz — CHANGES_REQUESTED** [review]

File: src/vcount.c [hunk 10]
```diff
@@ -98,3 +98,103 @@ void sub_08006FF0(void)
...
+// https://cexplore.karathan.at/z/eozt9J
+// the function seems to use lots of local variables.
+// I'm not sure how to force it to use r8, and I don't know how to use both r0 and r2 for gUnk_030031D0
+// also, the way to access gUnk_03003350 is weird. I can't get it to generate 'add r0, r0, #0x8'
+// I tried different local variable declarations but no use.
+// the link above is the close version, while the NONMATCHING version below does not use any local variable.
+
+#ifdef NONMATCHING
+
+void sub_0800708C(void)
```
Comment (jiangzhengwenjz) [inline src/vcount.c:130]: https://cexplore.karathan.at/z/OEIea3

**Review by jiangzhengwenjz — CHANGES_REQUESTED** [review]

File: asm/code_vcount.s (hunk 1 above)
Comment (Cofflag) [reply asm/code_vcount.s:7]: 708C is a nonmatching so I reserve it to create a scratch.

File: src/script.c (hunk 8 above)
Comment (Cofflag) [reply src/script.c:302]: I misunderstand the Strict Aliasing rule, sorry.

File: src/copy.c [hunk 11]
```diff
@@ -16,6 +16,103 @@ const u8 gUnk_080BAAF8[3][4] = {
...
+    It's weird that the if statement in this function gets ignored completely.
+    https://cexplore.karathan.at/z/rIt42N
+    Maybe just I made something wrong.
+*/
+
+#ifdef NONMATCHING
+
+u16 sub_08009238(u8 r0, u8 r1)
+{
+    u16 result;
```
Comment (jiangzhengwenjz) [inline src/copy.c:30]: https://cexplore.karathan.at/z/8HqnwQ
This is nowhere near hard but you made so many mistakes. I believe you can make more effort.

**Review by jiangzhengwenjz — CHANGES_REQUESTED** [review]

File: src/script.c (hunk 9 above)
Comment (Cofflag) [reply src/script.c:272]: No

Comment (Cofflag) [reply src/script.c:272]: I guess I will just create a scratch

File: src/script.c [hunk 12]
```diff
@@ -1,8 +1,16 @@
...
-#include "ramptr.h"
+#include "script.h"
 #include "data.h"
 #include "functions.h"
 
+extern u32 (*gUnk_08B716B4[])(void);
+extern u32 (*gUnk_08B716F4[])(void);
+extern u32 (*gUnk_08B71750[])(void);
+extern u32 (*gUnk_08B718B0[])(void);
+extern u32 (*gUnk_08B71AEC[])(void);
```
Comment (jiangzhengwenjz) [inline src/script.c:10]:
```suggestion
extern u32 (*const gUnk_08B716B4[])(void);
extern u32 (*const gUnk_08B716F4[])(void);
extern u32 (*const gUnk_08B71750[])(void);
extern u32 (*const gUnk_08B718B0[])(void);
extern u32 (*const gUnk_08B71AEC[])(void);
```

File: src/vcount.c [hunk 13]
```diff
@@ -98,3 +98,45 @@ void sub_08006FF0(void)
...
+    u32 r5 = gUnk_03003230;
+    struct Unk_030031D0 *r0, *r2;
+    u32 *r6;
+    u16 r1;
+    gUnk_03003360[r5] = 0;
+    gUnk_03003350[r5] = 0;
+    (gUnk_03003350 + 2)[r5] = 0;
+    gUnk_03003210 = 0;
+    r6 = gUnk_03003470;
+    r2 = &gUnk_030031D0[0][0];
```
Comment (jiangzhengwenjz) [inline src/vcount.c:132]:
```suggestion
    r2 = gUnk_030031D0[0];
```

File: src/vcount.c [hunk 14]
```diff
@@ -98,3 +98,45 @@ void sub_08006FF0(void)
...
+    r1 = 0xff;
+    r2 = r0 = &gUnk_030031D0[r5][0];
```
Comment (jiangzhengwenjz) [inline src/vcount.c:134]:
```suggestion
    r2 = r0 = gUnk_030031D0[r5];
```

**Review by jiangzhengwenjz — CHANGES_REQUESTED** [review]

**Review by jiangzhengwenjz — CHANGES_REQUESTED** [review]

File: src/attribute_O2_3.c [hunk 15]
```diff
@@ -31,7 +31,7 @@ void sub_08018EF8(u32 r4, u8 *r5, u32 *r6)
     u32 temp;
     for (i = 1; i <= 0x63; ++i)
     {
-        if ((i - 1) <= 0x62)
+        if ((u8)((u8)i - 1) <= 0x62)
```
Comment (jiangzhengwenjz) [inline src/attribute_O2_3.c:34]: This is just `i > 0 && i < 100`

**Review by jiangzhengwenjz — CHANGES_REQUESTED** [review]

File: src/attribute_O2_3.c [hunk 16]
```diff
@@ -19,3 +19,116 @@ u16 sub_08018EE8(u8 idx)
 {
     return gUnk_08B80178[idx];
 }
+
+// I am not sure how to fix the rest issues.
```
Comment (jiangzhengwenjz) [inline src/attribute_O2_3.c:23]: If you really have no idea, then send all the links to me. I don't want to scan all your functions here.

**Review by jiangzhengwenjz — CHANGES_REQUESTED** [review]

File: src/attribute_O2_3.c (hunk 16 above)
Comment (Cofflag) [reply src/attribute_O2_3.c:23]: Sorry. I will send the links.

File: src/attribute_O2_3.c [hunk 17]
```diff
@@ -19,3 +19,104 @@ u16 sub_08018EE8(u8 idx)
...
+}
+
+extern u32 sub_08016E58(u8 a);
+
+void sub_08018F64(u32 r6, u8 *r7, u32 *r8)
+{
+    u32 r4,temp;
+    for (r4 = 1; r4 <= 0x63; r4++)
+    {
+        if ((sub_08016E58(r4)) > r6) break;
```
Comment (jiangzhengwenjz) [inline src/attribute_O2_3.c:71]:
```suggestion
        if (sub_08016E58(r4) > r6) break;
```

File: src/attribute_O2_3.c [hunk 18]
```diff
@@ -19,3 +19,104 @@ u16 sub_08018EE8(u8 idx)
...
+        }
+        *r6 = temp - r4;
+    }
+}
+
+extern u32 sub_08016E58(u8 a);
+
+void sub_08018F64(u32 r6, u8 *r7, u32 *r8)
+{
+    u32 r4,temp;
```
Comment (jiangzhengwenjz) [inline src/attribute_O2_3.c:68]:
```suggestion
    u32 r4;
```

File: src/attribute_O2_3.c [hunk 19]
```diff
@@ -19,3 +19,104 @@ u16 sub_08018EE8(u8 idx)
...
+    for (r4 = 1; r4 <= 0x63; r4++)
```
Comment (jiangzhengwenjz) [inline src/attribute_O2_3.c:69]:
```suggestion
    for (r4 = 1; r4 < 100; r4++)
```

File: src/script.c [hunk 20]
```diff
@@ -146,3 +153,154 @@ u32 sub_08012F60(u16 r0)
...
+void sub_0801309C(void)
+{
+    gUnk_03006574->unk2 = *gUnk_03006574->unk4++;
+    sub_080127E4();
+    gUnk_03006578->unk1 = 1;
+}
+
+void sub_080130C4(void)
+{
+    if (gUnk_08B716B4[(u8)gUnk_03006574->unk2]() << 0x10 == 0)
```
Comment (jiangzhengwenjz) [inline src/script.c:212]: The left-shift is caused by 16-bit return value

File: src/script.c [hunk 21]
```diff
@@ -146,3 +153,154 @@ u32 sub_08012F60(u16 r0)
...
+    {
+        gUnk_03006574->unk2 = *gUnk_03006574->unk4++;
+        sub_080127E4();
+    }
+}
+
+void sub_080130FC(void)
+{
+    if (gUnk_08B716F4[(u8)gUnk_03006574->unk2]() << 0x10 == 0)
```
Comment (jiangzhengwenjz) [inline src/script.c:221]: The left-shift is caused by 16-bit return value

File: src/script.c [hunk 22]
```diff
@@ -146,3 +153,154 @@ u32 sub_08012F60(u16 r0)
...
+    {
+        gUnk_03006574->unk2 = *gUnk_03006574->unk4++;
+        sub_080127E4();
+    }
+}
+
+void sub_08013134(void)
+{
+    if (gUnk_08B71750[(u8)gUnk_03006574->unk2]() << 0x10 == 0)
```
Comment (jiangzhengwenjz) [inline src/script.c:230]: The left-shift is caused by 16-bit return value

File: src/script.c [hunk 23]
```diff
@@ -146,3 +153,154 @@ u32 sub_08012F60(u16 r0)
...
+    {
+        gUnk_03006574->unk2 = *gUnk_03006574->unk4++;
+        sub_080127E4();
+    }
+}
+
+void sub_0801316C(void)
+{
+    if (gUnk_08B718B0[(u8)gUnk_03006574->unk2]() << 0x10 == 0)
```
Comment (jiangzhengwenjz) [inline src/script.c:239]: The left-shift is caused by 16-bit return value

File: src/script.c [hunk 24]
```diff
@@ -146,3 +153,154 @@ u32 sub_08012F60(u16 r0)
...
+    {
+        gUnk_03006574->unk2 = *gUnk_03006574->unk4++;
+        sub_080127E4();
+    }
+}
+
+void sub_080131A4(void)
+{
+    if (gUnk_08B71AEC[(u8)gUnk_03006574->unk2]() << 0x10 == 0)
```
Comment (jiangzhengwenjz) [inline src/script.c:248]: The left-shift is caused by 16-bit return value

File: src/script.c [hunk 25]
```diff
@@ -146,3 +153,154 @@ u32 sub_08012F60(u16 r0)
...
+    {
+        gUnk_03006574->unk2 = *gUnk_03006574->unk4++;
+        sub_080127E4();
+    }
+}
+
+// https://cexplore.karathan.at/z/rhnv92
+
+u32 sub_080131DC(void)
```
Comment (jiangzhengwenjz) [inline src/script.c:257]: Please update the function

File: src/small_strucures.c [hunk 26]
```diff
@@ -502,3 +503,31 @@ void sub_08063E64(struct Unk_03006AF4 *r0)
...
     sub_08011104(10);
 }
+
+extern void (*gUnk_08BC9C68[])(s16 a, s16 b, s16 c, s16 d);
+extern s16 (*gUnk_08BC9C7C[])(void);
+extern u16 sub_08068070(u16);
+
+s16 sub_08063F70(void)
+{
+    u16 r3,r4,r5,r6;
```
Comment (jiangzhengwenjz) [inline src/small_strucures.c:513]:
```suggestion
    u16 r3, r4, r5, r6;
```

File: src/small_strucures.c [hunk 27]
```diff
@@ -502,3 +503,31 @@ void sub_08063E64(struct Unk_03006AF4 *r0)
...
+    struct Unk_03006AF4 *ptr = gUnk_03006AF4;
+    switch ((s16)ptr->unk2) // TODO:figure out why this weird cast keeps to happen
+    {
+        case 0 :
+            r6 = sub_08068070(1);
+            r5 = sub_08068070(2);
+            r4 = sub_08068070(3);
+            r3 = sub_08068070(4);
+            gUnk_08BC9C68[ptr[1].unk2](r6, r5, r4, r3);
+            ptr->unk2++; // tailmerge
```
Comment (jiangzhengwenjz) [inline src/small_strucures.c:523]: There's no point leaving the comment here.

File: src/small_strucures.c [hunk 28]
```diff
@@ -502,3 +503,31 @@ void sub_08063E64(struct Unk_03006AF4 *r0)
...
+
+extern void (*gUnk_08BC9C68[])(s16 a, s16 b, s16 c, s16 d);
+extern s16 (*gUnk_08BC9C7C[])(void);
+extern u16 sub_08068070(u16);
+
+s16 sub_08063F70(void)
+{
+    u16 r3,r4,r5,r6;
+    struct Unk_03006AF4 *ptr = gUnk_03006AF4;
+    switch ((s16)ptr->unk2) // TODO:figure out why this weird cast keeps to happen
```
Comment (jiangzhengwenjz) [inline src/small_strucures.c:515]: Maybe it's really s16?

File: src/small_strucures.c [hunk 29]
```diff
@@ -502,3 +503,31 @@ void sub_08063E64(struct Unk_03006AF4 *r0)
...
+            gUnk_08BC9C68[ptr[1].unk2](r6, r5, r4, r3);
+            ptr->unk2++; // tailmerge
+            break;
+        case 1 :
+            if (gUnk_08BC9C7C[ptr[1].unk2]() == 0)
+            {
+                ptr->unk2 = 0;
+            }
+            break;
+    }
```
Comment (jiangzhengwenjz) [inline src/small_strucures.c:531]: When coding a switch statement, please follow the coding style:
```suggestion
    switch ((s16)ptr->unk2) // TODO: figure out why this weird cast keeps to happen
    {
    case 0:
        r6 = sub_08068070(1);
        r5 = sub_08068070(2);
        r4 = sub_08068070(3);
        r3 = sub_08068070(4);
        gUnk_08BC9C68[ptr[1].unk2](r6, r5, r4, r3);
        ptr->unk2++;
        break;
    case 1:
        if (gUnk_08BC9C7C[ptr[1].unk2]() == 0)
        {
            ptr->unk2 = 0;
        }
        break;
    }
```

**Review by jiangzhengwenjz — CHANGES_REQUESTED** [review]

File: src/small_strucures.c (hunk 28 above)
Comment (Cofflag) [reply src/small_strucures.c:515]: It's not matching with s16😳

Comment (jiangzhengwenjz) [reply src/small_strucures.c]: > It's not matching with s16😳

Where?

Comment (Cofflag) [reply src/small_strucures.c]: https://cexplore.karathan.at/z/c2Za1u
I think it's the function array part.

Comment (jiangzhengwenjz) [reply src/small_strucures.c]: Then you can cast it to u16 instead. s16 -> u16 is implementation-defined, but s16 -> u16 is always well-defined.

**Comment by jiangzhengwenjz** [comment]
```
757684 total bytes of code
20960 bytes of code in src (2.7663%)
736724 bytes of code in asm (97.2337%)
```

**Pull request merged**